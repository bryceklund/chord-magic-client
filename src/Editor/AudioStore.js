const { API_BASE_URL, API_TOKEN } = require('../config')
const chordsUrl = `${API_BASE_URL}/chords`
const scalesUrl = `${API_BASE_URL}/scales`
const options = {
    method: 'GET',
    headers: new Headers({
      'Authorization': `Bearer ${API_TOKEN}`,
      'Content-Type': 'application/json'
    })
  }

let chords
let scales
let final = {}

fetch(chordsUrl, options)
  .then(data => data.json())
  .then(data => chords = data)
  .then(data => {
    fetch(scalesUrl, options)
        .then(data => data.json())
        .then(data => scales = data)
        .then(data => {
            scales.map((scale, i) => {
                final[scale] = []
            })
        })
        .then(data => {
            scales.forEach((scale, i) => {
                chords.map((chord, i) => {
                    fetch(`${API_BASE_URL}/chord/${scale}/${chord}`, options)
                        .then(data => data.json())
                        .then(data => final[scale][chord] = data)
                })
            })
        })
        .then(data => AudioStore.chords = final)
  })


const AudioStore = {
    notes: {
        minusTwo: {
            A: 110.00,
            Bb: 116.54,
            B: 123.47,
            C: 130.81,
            Db: 138.59,
            D: 146.83,
            Eb: 155.56,
            E: 164.81,
            F: 174.61,
            Gb: 185.00,
            G: 196.00,
            Ab: 207.65
        },
        minusOne: {
            A: 220.00,
            Bb: 233.08,
            B: 246.94,
            C: 261.63,
            Db: 277.18,
            D: 293.66,
            Eb: 311.13,
            E: 329.63,
            F: 349.23,
            Gb: 369.99,
            G: 392.00,
            Ab: 415.30
        },
        zero: {    
            A: 440,
            Bb: 466.16,
            B: 493.88,
            C: 523.25,
            Db: 554.37,
            D: 587.33,
            Eb: 622.25,
            E: 659.25,
            F: 698.46,
            Gb: 739.99,
            G: 783.99,
            Ab: 830.61
        },
        plusOne: {
            A: 880.00,
            Bb: 932.33,
            B: 987.77,
            C: 1046.50,
            Db: 1108.73,
            D: 1174.66,
            Eb: 1244.51,
            E: 1318.51,
            F: 1396.91,
            Gb: 1479.98,
            G: 1567.98,
            Ab: 1661.22
        },
        plusTwo: {
            A: 1760.00,
            Bb: 1864.66,
            B: 1975.53,
            C: 2093.00,
            Db: 2217.46,
            D: 2349.32,
            Eb: 2489.02,
            E: 2637.02,
            F: 2793.83,
            Gb: 2959.96,
            G: 3135.96,
            Ab: 3322.44
        }
    },
    chords: {}
}

module.exports = AudioStore