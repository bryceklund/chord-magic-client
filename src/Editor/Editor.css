.response {
    position: absolute;
    text-align: center;
}

.success {
    color: green;
}

.failure {
    color: red;
}