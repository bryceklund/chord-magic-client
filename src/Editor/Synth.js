import unmute from './unmute'

const AudioStore = require('./AudioStore.js')
const AudioUnlock = require('web-audio-touch-unlock')

function Synth(voice, oct, dur, vol, scale, chordName) {
    //create instance of the API
    const context = new (window.AudioContext || window.webkitAudioContext)()

    //notes in this octave
    const notes = AudioStore.notes[oct]

    //notes in this chord
    const chord = AudioStore.chords[scale][chordName]

    console.log(notes, chord)
    //create enough oscillators
    let synth = []
    for (let i = 0; i < chord.length; i++) {
        synth[i] = context.createOscillator()
    }

    //set voice types
    //eg. 'triangle', 'saw', 'square', or 'sine'
    synth.map((osc) => {
        osc.type = voice
    })

    //set osc notes
    synth.map((osc, i) => {
        osc.frequency.value = notes[chord[i]]
    })

    //create gain instance 
    const volume = context.createGain()
    volume.gain.value = vol

    //connect gain to computer audio
    volume.connect(context.destination)

    //connect oscillators to gain instance
    synth.map(osc => {
        osc.connect(volume)
    })

    //get current time from API
    const startTime = context.currentTime

    //fade in and out
    volume.gain.setValueAtTime(0, startTime)
    volume.gain.linearRampToValueAtTime(vol, startTime + (dur / 10))
    volume.gain.exponentialRampToValueAtTime(0.01, startTime + dur)

    unmute(context)

    //start and stop playback
    synth.map(osc => {
        osc.start(startTime)
        osc.stop(startTime + dur)
    })
}

export default Synth;
