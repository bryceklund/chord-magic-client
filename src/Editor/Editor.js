import React, { Component } from 'react'
import Timeline from './Timeline/Timeline.js'
import Library from './Library/Library.js'
import { API_BASE_URL, API_TOKEN } from '../config'
import './Editor.css'
import TokenService from '../services/tokenService.js'

class Editor extends Component {
    constructor(props) {
        super(props)
        this.timelineElement = React.createRef()
        this.libraryElement = React.createRef()
        this.state = {
            librarySelection: false,
            timelineSelection: false,
            selectedChord: {},
            selectedIndex: null,
            response: ''
        }
    }

    refresh = () => {
        this.forceUpdate()
    }

    loadProgression = () => {
        this.timelineElement.current.loadProgression(this.props.progression)
    }

    setChordIndex = (selectedIndex) => {
        this.setState({
            selectedIndex
        })
    }

    insertChord = () => {
        this.timelineElement.current.insertChord(this.state.selectedIndex, this.state.selectedChord)
        this.libraryElement.current.resetChordSelection()
        this.setState({
            selectedChord: {},
            selectedIndex: null,
            librarySelection: false
        })
    }

    toggleLibrarySelection = (active) => {
        if (active) {
            this.setState({
                librarySelection: true
            })
        } else {
            this.setState({
                librarySelection: false
            })
        }
    }

    toggleTimelineSelection = (active) => {
        if (active) {
            this.setState({
                timelineSelection: true
            })
        } else {
            this.setState({
                timelineSelection: false
            })
        }
    }

    updateSelectedChord = (voice, oct, scale, name) => {
        this.setState({
            selectedChord: {
                name,
                scale,
                oct,
                voice
            }
        })
    }

    clearSelectedChord = () => {
        this.setState({
            selectedChord: {}
        })
    }

    saveNew = (name) => {
        const chords = this.timelineElement.current.state.activeChords.map((chord, i) => chord = {...chord, id: i})
        const url = `${API_BASE_URL}/progressions/saved`
        const options = {
          method: 'POST',
          headers: new Headers({
            'Authorization': `Bearer ${API_TOKEN} ${TokenService.getAuthToken()}`,
            'Content-Type': 'application/json'
          }),
          body: JSON.stringify({
            name: name,
            chords
          })
        }
        let id
        //post to db
        fetch(url, options)
            .then(res => res.text())
            .then(res => {
                if (res.length) {
                    this.libraryElement.current.saveSuccess()
                    id = res
                } else {
                    this.libraryElement.current.saveFail()
                }
            })
            .then(res => this.props.setCurrentProg(name, id))
      }

      saveExisting = () => {
        const chords = this.timelineElement.current.state.activeChords.map((chord, i) => chord = {...chord, id: i})
        const id = this.props.currentId
        const url = `${API_BASE_URL}/progressionchords/${id}`
        const options = {
          method: 'PATCH',
          headers: new Headers({
            'Authorization': `Bearer ${API_TOKEN}`,
            'Content-Type': 'application/json'
          }),
          body: JSON.stringify({
            chords
          })
        }
        fetch(url, options)
            .then(res => {
                if (res.ok) {
                    this.libraryElement.current.saveSuccess()
                } else {
                    this.libraryElement.current.saveFail()
                }
            })
            
      }

    render() {
        return (
            <React.Fragment>
                <Timeline ref={this.timelineElement} currentName={this.props.currentName} progression={this.props.progression} chordSelected={this.toggleTimelineSelection} setIndex={this.setChordIndex} />
                <Library 
                    setProgName={this.props.setProgName} 
                    signedIn={this.props.signedIn} 
                    saveNew={this.saveNew}
                    saveExisting={this.saveExisting}
                    currentName={this.props.currentName} 
                    selected={this.state.librarySelection} 
                    ref={this.libraryElement} 
                    insertChord={this.insertChord} 
                    chordSelected={this.toggleLibrarySelection} 
                    storeChord={this.updateSelectedChord} 
                    resetSelection={this.clearSelectedChord} />
            </React.Fragment>
        );
    }
}

export default Editor;