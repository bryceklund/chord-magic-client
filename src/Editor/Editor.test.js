import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import { BrowserRouter } from 'react-router-dom'
import Editor from './Editor'

describe(`Editor`, () => {
    it(`renders without crashing`, () => {
        const div = document.createElement('div')
        ReactDOM.render(<BrowserRouter>
                            <Editor />
                        </BrowserRouter>, div)
        ReactDOM.unmountComponentAtNode(div)
    })
    it(`matches its snapshot`, () => {
        const tree = renderer.create(<BrowserRouter>
                                        <Editor />
                                    </BrowserRouter>)
                                    .toJSON()
        expect(tree).toMatchSnapshot()
    })
})
