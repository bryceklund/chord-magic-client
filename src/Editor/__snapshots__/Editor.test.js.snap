// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editor matches its snapshot 1`] = `
Array [
  <section
    className="timeline"
  >
    <div
      className="hidden"
    >
      now editing: "
      "
    </div>
    <div
      className="hidden"
    >
      <div
        className="note_container"
        viewBox="25 25 50 50"
      >
        <img
          className="music_note"
          src="loading-icon.png"
        />
      </div>
      Oops, try again in a few seconds...
    </div>
    <div
      className="play_stop_buttons"
    >
      <button
        onClick={[Function]}
      >
        <img
          alt="play button"
          className="play_stop_icon"
          src="play-button.png"
        />
      </button>
          
      <button
        onClick={[Function]}
      >
        <img
          alt="stop button"
          className="play_stop_icon"
          src="stop-button.png"
        />
      </button>
    </div>
    <div
      className="global_view"
    >
      <span
        className="global maj g"
      >
        _
      </span>
      <span
        className="global maj b"
      >
        _
      </span>
      <span
        className="global min e"
      >
        _
      </span>
      <span
        className="global maj c"
      >
        _
      </span>
      <span
        className="global_inactive g"
      >
        _
      </span>
      <span
        className="global_inactive b"
      >
        _
      </span>
      <span
        className="global_inactive c"
      >
        _
      </span>
    </div>
    <div
      className="player_timeline"
    >
      <div
        className="chord_box one maj g"
      >
        <button
          className="delete_chord"
          onClick={[Function]}
        >
          <img
            alt="delete chord"
            src="delete-button.png"
          />
        </button>
        <span
          className="chord_label "
          onClick={[Function]}
        >
          G
          <section
            className="chord_info hidden"
          >
            <div
              className="info_label"
            >
              maj
            </div>
            <div
              className="info_label"
            >
              sawtooth
            </div>
            <div
              className="info_label"
            >
              -2
            </div>
          </section>
        </span>
        <div
          className="move_chord_buttons"
        >
          <button
            className="move_chord_right"
            onClick={[Function]}
          >
            <img
              alt="move right"
              src="move-right-arrow.png"
            />
          </button>
        </div>
      </div>
      <div
        className="chord_box two maj b"
      >
        <button
          className="delete_chord"
          onClick={[Function]}
        >
          <img
            alt="delete chord"
            src="delete-button.png"
          />
        </button>
        <span
          className="chord_label "
          onClick={[Function]}
        >
          B
          <section
            className="chord_info hidden"
          >
            <div
              className="info_label"
            >
              maj
            </div>
            <div
              className="info_label"
            >
              sawtooth
            </div>
            <div
              className="info_label"
            >
              -2
            </div>
          </section>
        </span>
        <div
          className="move_chord_buttons"
        >
          <button
            className="move_chord_left"
            onClick={[Function]}
          >
            <img
              alt="move left"
              src="move-left-arrow.png"
            />
          </button>
          <button
            className="move_chord_right"
            onClick={[Function]}
          >
            <img
              alt="move right"
              src="move-right-arrow.png"
            />
          </button>
        </div>
      </div>
      <div
        className="chord_box three min e"
      >
        <button
          className="delete_chord"
          onClick={[Function]}
        >
          <img
            alt="delete chord"
            src="delete-button.png"
          />
        </button>
        <span
          className="chord_label "
          onClick={[Function]}
        >
          E
          <section
            className="chord_info hidden"
          >
            <div
              className="info_label"
            >
              min
            </div>
            <div
              className="info_label"
            >
              sawtooth
            </div>
            <div
              className="info_label"
            >
              -2
            </div>
          </section>
        </span>
        <div
          className="move_chord_buttons"
        >
          <button
            className="move_chord_left"
            onClick={[Function]}
          >
            <img
              alt="move left"
              src="move-left-arrow.png"
            />
          </button>
          <button
            className="move_chord_right"
            onClick={[Function]}
          >
            <img
              alt="move right"
              src="move-right-arrow.png"
            />
          </button>
        </div>
      </div>
      <div
        className="chord_box four maj c"
      >
        <button
          className="delete_chord"
          onClick={[Function]}
        >
          <img
            alt="delete chord"
            src="delete-button.png"
          />
        </button>
        <span
          className="chord_label "
          onClick={[Function]}
        >
          C
          <section
            className="chord_info hidden"
          >
            <div
              className="info_label"
            >
              maj
            </div>
            <div
              className="info_label"
            >
              sawtooth
            </div>
            <div
              className="info_label"
            >
              -2
            </div>
          </section>
        </span>
        <div
          className="move_chord_buttons"
        >
          <button
            className="move_chord_left"
            onClick={[Function]}
          >
            <img
              alt="move left"
              src="move-left-arrow.png"
            />
          </button>
          <button
            className="move_chord_right"
            onClick={[Function]}
          >
            <img
              alt="move right"
              src="move-right-arrow.png"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      className="scrubber_buttons"
    >
      <button
        onClick={[Function]}
      >
        <img
          alt="scrub left"
          src="scrub-left-arrow.png"
        />
      </button>
      <div
        className="slide_container"
      >
        <label
          className="volume"
          htmlFor="volume"
        >
          Volume: 
          25
          %
        </label>
        <input
          className="volume_slider"
          defaultValue="25"
          id="volume"
          max="100"
          min="0"
          onChange={[Function]}
          type="range"
        />
      </div>
      <div
        className="slide_container"
      >
        <label
          className="speed"
          htmlFor="speed"
        >
          Speed: 
          160
           BPM
        </label>
        <input
          className="speed_slider"
          defaultValue="30"
          id="speed"
          max="100"
          min="1"
          onChange={[Function]}
          type="range"
        />
      </div>
      <button
        onClick={[Function]}
      >
        <img
          alt="scrub right"
          src="scrub-right-arrow.png"
        />
      </button>
    </div>
    <div
      className="jump_buttons"
    >
      <button
        onClick={[Function]}
      >
        <img
          alt="jump to beginning"
          src="jump-left-arrow.png"
        />
      </button>
      <button
        onClick={[Function]}
      >
        <img
          alt="jump to end"
          src="jump-right-arrow.png"
        />
      </button>
    </div>
  </section>,
  <section
    className="library_container"
  >
    <div
      className="library_settings"
    >
      <div
        className="settings"
      >
        <div
          className="oct_ins_pre"
        >
          <div
            className="prehear_container"
          >
            <input
              className="prehear"
              id="prehear"
              onChange={[Function]}
              type="checkbox"
            />
            <label
              className="prehear_label"
              htmlFor="prehear"
            >
              prehear
            </label>
          </div>
          <div
            className="instrument_container"
          >
            instrument
            <div
              className="wave_icons_container"
            >
              <div
                className="sine_triangle"
              >
                <label>
                  <input
                    name="voice"
                    onClick={[Function]}
                    type="radio"
                    value="sine"
                  />
                  <img
                    className="wave_icon"
                    src="sine.png"
                  />
                </label>
                <label>
                  <input
                    defaultChecked={true}
                    name="voice"
                    onClick={[Function]}
                    type="radio"
                    value="triangle"
                  />
                  <img
                    className="wave_icon"
                    src="triangle.png"
                  />
                </label>
              </div>
              <div
                className="saw_square"
              >
                <label>
                  <input
                    name="voice"
                    onClick={[Function]}
                    type="radio"
                    value="sawtooth"
                  />
                  <img
                    className="wave_icon"
                    src="sawtooth.png"
                  />
                </label>
                <label>
                  <input
                    name="voice"
                    onClick={[Function]}
                    type="radio"
                    value="square"
                  />
                  <img
                    className="wave_icon"
                    src="square.png"
                  />
                </label>
              </div>
            </div>
          </div>
          <div
            className="octave_container"
          >
            <label
              htmlFor="octave"
            >
              octave: 
              0
            </label>
            <br />
            <input
              className="octave"
              defaultValue="0"
              max="2"
              min="-2"
              name="octave"
              onChange={[Function]}
              type="range"
            />
          </div>
        </div>
        <button
          className="insert_chord"
          disabled={true}
          onClick={[Function]}
        >
          insert chord
        </button>
        <p
          className="login_message"
        >
          Log in to save your progression!
        </p>
        <div
          className="new_save_tooltip response hidden"
          id="response"
        >
          
        </div>
        <div
          className="update_tooltip hidden"
          id="update_tooltip"
        >
          <button
            className="save"
            onClick={[Function]}
          >
            Save as new
          </button>
          <button
            className="save"
            onClick={[Function]}
          >
            Overwrite existing
          </button>
          <button
            className="save"
            onClick={[Function]}
          >
            Cancel
          </button>
        </div>
        <div
          className="new_save_tooltip hidden"
          id="new_save_tooltip"
        >
          <label
            htmlFor="prog_name"
          >
            Name:
          </label>
          <input
            id="prog_name"
            onChange={[Function]}
            type="text"
          />
          <div
            className="save_cancel"
          >
            <button
              className="save"
              onClick={[Function]}
            >
              Save
            </button>
            <button
              className="save"
              onClick={[Function]}
            >
              Cancel
            </button>
          </div>
        </div>
      </div>
      <div
        className="library"
      >
        <button
          className="back hidden"
          onClick={[Function]}
        >
          &lt; back to scales
        </button>
        <ul
          className="scales "
        >
          <li
            className="scale maj"
            onClick={[Function]}
          >
            Maj
          </li>
          <li
            className="scale min"
            onClick={[Function]}
          >
            Min
          </li>
          <li
            className="scale maj7"
            onClick={[Function]}
          >
            Maj7
          </li>
          <li
            className="scale min7"
            onClick={[Function]}
          >
            Min7
          </li>
          <li
            className="scale dim"
            onClick={[Function]}
          >
            Dim
          </li>
          <li
            className="scale dom7"
            onClick={[Function]}
          >
            Dom7
          </li>
          <li
            className="scale sus2"
            onClick={[Function]}
          >
            Sus2
          </li>
          <li
            className="scale sus4"
            onClick={[Function]}
          >
            Sus4
          </li>
          <li
            className="scale aug"
            onClick={[Function]}
          >
            Aug
          </li>
        </ul>
        <ul
          className="chords hidden"
        >
          <li
            className="note a "
            onClick={[Function]}
          >
            A
          </li>
          <li
            className="note bb "
            onClick={[Function]}
          >
            Bb
          </li>
          <li
            className="note b "
            onClick={[Function]}
          >
            B
          </li>
          <li
            className="note c "
            onClick={[Function]}
          >
            C
          </li>
          <li
            className="note db "
            onClick={[Function]}
          >
            Db
          </li>
          <li
            className="note d "
            onClick={[Function]}
          >
            D
          </li>
          <li
            className="note eb "
            onClick={[Function]}
          >
            Eb
          </li>
          <li
            className="note e "
            onClick={[Function]}
          >
            E
          </li>
          <li
            className="note f "
            onClick={[Function]}
          >
            F
          </li>
          <li
            className="note gb "
            onClick={[Function]}
          >
            Gb
          </li>
          <li
            className="note g "
            onClick={[Function]}
          >
            G
          </li>
          <li
            className="note ab "
            onClick={[Function]}
          >
            Ab
          </li>
        </ul>
      </div>
    </div>
  </section>,
]
`;
